File: 23.Merge-k-Sorted-Lists/list.py
from typing import Optional
from heapq import heapify, heappush, heappop


class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next


class Solution:
    def mergeKLists(self, lists: list[Optional[ListNode]]) -> Optional[ListNode]:
        # O(n)
        idxs = [(node.val, i) for i, node in enumerate(lists) if node]

        # O(n)
        heapify(idxs)

        curr = None
        root = None

        # O(Nlogn)
        while idxs:
            _, i = heappop(idxs)
            if not root:
                root = lists[i]

            if curr:
                curr.next = lists[i]
            curr = lists[i]

            if next := lists[i].next:
                lists[i] = next
                heappush(idxs, (next.val, i))

        return root

File: 23.Merge-k-Sorted-Lists/list.rs
use std::{
    cmp::{Ordering, Reverse},
    collections::BinaryHeap,
};

impl PartialOrd for ListNode {
    fn partial_cmp(&self, b: &ListNode) -> Option<Ordering> {
        Some(self.cmp(b))
    }
}

impl Ord for ListNode {
    fn cmp(&self, b: &ListNode) -> Ordering {
        self.val.cmp(&b.val)
    }
}

impl Solution {
    pub fn merge_k_lists(lists: Vec<Option<Box<ListNode>>>) -> Option<Box<ListNode>> {
        // O(n)
        let mut nodes: BinaryHeap<_> = lists.into_iter().flatten().map(Reverse).collect();
        let mut head = ListNode::new(0);
        let mut tail = &mut head;

        // O(Nlogn)
        while let Some(Reverse(mut node)) = nodes.pop() {
            if let Some(next) = node.next.take() {
                nodes.push(Reverse(next));
            }

            tail = tail.next.insert(node);
        }

        head.next
    }
}
